sip/sip-006-runtime-cost-assessment.md

```diff
@@ -0,0 +1,625 @@
+or statically. Dynamic cost assessment involves tracking, at the VM level,
+the various metrics as a contract is executed. Static cost assessment is
+performed via analysis of the contract source code, and is inherently
+a more pessimistic accounting of the execution cost: list operations
+are charged according to the _maximum_ size of the list (per the type
+annotations and inferences from the source code) and branching statements
+are charged according to the _maximum_ branch cost (per metric tracked, i.e.,
+if one branch performs 1 write and has a runtime cost of 1, and another
+branch performs 0 writes and has a runtime cost of 2, the whole statement
+will be assessed as having a maximum of 1 write and runtime cost of 2).
+
+# Costs of Common Operations
+
+### Variable Lookup
+
+Looking up variables in Clarity incurs a non-constant cost -- the stack
+depth _and_ the length of the variable name affect this cost. However,
+variable names in Clarity have bounded length -- 128 characters. Therefore,
+the cost assessed for variable lookups may safely be constant with respect
+to name length.
+
+The stack depth affects the lookup cost because the variable must be
+checked for in each context on the stack.
+
+The cost model of Clarity depends on a copy-on-read semantic for
+objects. This allows operations like appends, matches, wrapping/unwrapping,
+to be constant cost, but it requires that variable lookups be charged for
+copies.
+
+Cost Function:
+
+```
+a*X+b*Y+c
+```
+
+where a, b, and c are constants,
+X := stack depth
+Y := variable size
+
+### Function Lookup
+
+Looking up a function in Clarity incurs a constant cost with respect
+to name length (for the same reason as variable lookup). However,
+because functions may only be defined in the top-level contract
+context, stack depth does not affect function lookup.
+
+Cost Function:
+
+```
+a
+```
+
+where a is a constant.
+
+### Name Binding
+
+The cost of binding a name in Clarity -- in either a local or the contract
+context is _constant_ with respect to the length of the name:
+
+```
+binding_cost = a
+```
+
+where a is a constant
+
+### Function Application
+
+Function application in Clarity incurs a cost in addition to the
+cost of executing the function's body. This cost is the cost of
+binding the arguments to their passed values, and the cost of
+ensuring that those arguments are of the correct type. Type checks
+and argument binding are _linear_ in the size of the arguments.
+
+The cost of applying a function is:
+
+
+```
+(a*X+b) + costEval(body)
+```
+
+where a and b are constants,
+X := the cumulative size of the argument types,
+costEval(body) := the cost of executing the body of the function
+
+### Type Parsing
+
+Parsing a type in Clarity incurs a linear cost in the size of the
+AST describing the type:
+
+```
+type_parsing_cost(X) = (a*X+b)
+```
+
+where a, b, are constants,
+X := the number of elements in the type description AST
```

Comment:
Do you think you can list a couple of examples of what a `type description AST` is, and how it's different from the program AST?

Reply:
Sure -- added in 9613e137e